#  Excel Handler for Data Analysis

## Kohort Data Analysis

In [16]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split

In [17]:
# Load the Excel file
file_path = '어린이코호트 임상데이터.xlsx' 
df = pd.read_excel(file_path)
# Display the first few rows of the dataset
print(df.head())

   환자명 성별  나이  ALB(S)  ALT  AST  Iron(S)  Free T4   T3    TSH  ...   WBC  RBC  \
0    1  여   6     4.7    9   27       99     1.27  135  1.865  ...  6.43  4.4   
1    2  여   7     4.8   40   32      147     1.45  158  1.344  ...  6.37  4.3   
2    3  여   6     4.8    9   28      115     1.34  115  2.405  ...  5.62  4.4   
3    4  여   7     4.7   16   29      109     1.20  132  1.223  ...  6.32  4.3   
4    5  여   7     4.7    9   27      116     1.21  152  2.346  ...  4.53  4.0   

  Hemoglobin  Hematocrit  Platelet  HbA1c  FSH      E2  Testosterone     LH  
0       13.2        38.7       317    5.5  2.6  < 5.00           NaN  < 0.3  
1       12.7        36.5       368    5.3  1.7       5           NaN    0.3  
2       12.0        37.8       283    5.4  1.6       5           NaN    0.3  
3       12.5        36.3       336    5.2  1.2  < 5.00           NaN  < 0.3  
4       11.4        33.7       277    5.5  2.1       5           NaN    0.3  

[5 rows x 23 columns]


In [18]:
df.describe()

,환자명,나이,ALB(S),ALT,AST,Iron(S),Free T4,T3,TSH,T.Cholestero,WBC,RBC,Hemoglobin,Hematocrit,Platelet,HbA1c,Testosterone
count,313.00000,313.000000,313.000000,313.000000,313.000000,313.000000,313.000000,313.000000,313.000000,313.000000,313.000000,313.000000,313.000000,313.000000,313.000000,313.000000,0.0
mean,157.00000,7.805112,4.789457,13.290735,25.763578,94.290735,1.327316,159.472843,2.508987,175.031949,6.362428,4.594249,13.366773,40.276358,309.370607,5.351438,NaN
std,90.49954,1.449040,0.227551,6.169752,7.287262,31.274454,0.191620,25.044065,1.301915,28.779649,1.752727,0.273449,0.769395,2.455596,68.923104,0.256342,NaN
min,1.00000,6.000000,3.500000,3.000000,16.000000,28.000000,0.950000,97.000000,0.470000,103.000000,2.290000,3.800000,11.100000,32.200000,156.000000,4.400000,NaN
25%,79.00000,7.000000,4.700000,10.000000,22.000000,73.000000,1.210000,143.000000,1.560000,155.000000,5.260000,4.400000,12.900000,38.500000,262.000000,5.200000,NaN
50%,157.00000,7.000000,4.800000,12.000000,25.000000,93.000000,1.320000,158.000000,2.270000,173.000000,6.230000,4.600000,13.300000,40.400000,302.000000,5.300000,NaN
75%,235.00000,9.000000,4.900000,15.000000,28.000000,113.000000,1.430000,175.000000,3.090000,192.000000,7.320000,4.800000,13.900000,42.000000,346.000000,5.500000,NaN
max,313.00000,10.000000,5.700000,47.000000,127.000000,200.000000,3.310000,275.000000,7.430000,360.000000,12.250000,5.300000,15.300000,46.700000,575.000000,6.200000,NaN


📊 기본 통계 요약
🔹 연령 및 간/영양 관련

나이 평균: 7.8세
ALB(S): 평균 4.79
ALT: 평균 13.29
AST: 평균 25.76

👉 전반적으로 간수치는 소아 평균 범위 안에서 안정적인 분포

🔹 철분 및 갑상선 관련

Iron(S): 평균 94.29
Free T4: 평균 1.33
T3: 평균 159.47
TSH: 평균 2.51

👉 갑상선 수치는 비교적 정상 범위 중심
👉 일부 최대값(T3=275, TSH=7.43)은 outlier 가능성

🔹 혈액(CBC)

WBC: 평균 6.36
RBC: 평균 4.59
Hemoglobin: 평균 13.37 g/dL
Hematocrit: 평균 40.28%
Platelet: 평균 309

👉 (Sheet10 기준) 정상 범위에 전반적으로 잘 맞음

🔹 당뇨 관련

HbA1c 평균: 5.35%

👉 대부분 정상 범위 (4~6%) 내 분포

✅ 3. 데이터 특성 분석 (핵심 인사이트)
✔ 1) 분포 특징

대부분 변수는 정규분포에 가까운 구조
Platelet, T3, TSH는 right-skew (상위 이상치 존재)


✔ 2) 변동성 (표준편차 기준)

비교적 변동 큰 변수:

Iron (~31)
T3 (~25)
Platelet (~69)


안정적인 변수:

ALB (~0.22)
HbA1c (~0.25)




✔ 3) 이상치 가능 영역

ALT max = 47
AST max = 127
Platelet max = 575
TSH max = 7.43

👉 일부 개별 환자에서 임상적으로 의미 있는 outlier 존재 가능

✅ 4. 추가 분석 제안 (실무적으로 매우 중요)
원하시면 바로 이어서 도와드릴 수 있습니다:
📌 추천 분석

이상치 탐지 (IQR / Z-score)
나이별 그룹 비교 (6,7,8,9,10세)
상관분석 (예: Iron ↔ Hb)
질병 예측 feature 탐색
공복 여부 처리 (공복X 값 정리)

🔹 특이사항

Testosterone: 값이 거의 없음 (NaN)
일부 호르몬(E2, LH)은 텍스트 포함 (“< 5.00”) → 숫자형 분석 제한

✅ 결론

데이터는 잘 구조화되어 있고 통계적으로 안정적
일부 변수에 텍스트/결측치 존재 → 전처리 필요
소아 정상 범위 기준으로 보면 대부분 정상 분포


👉 다음 단계로 **“이상치 자동 검출 + 시각화(boxplot)”**까지 해드릴까요?

✅ 한눈에 정리 (구조 이해)
👉 데이터는 크게 5가지 영역으로 나뉩니다:

기본정보 (ID, 성별, 나이)
간/영양 (ALB, ALT, AST, Iron)
갑상선 (FreeT4, T3, TSH)
대사/당뇨 (콜레스테롤, 혈당, 인슐린, HbA1c)
혈액(CBC) + 호르몬

✅ 추가 팁 (분석할 때 중요)

In [19]:
liver = ['ALB(S)', 'ALT', 'AST']
thyroid = ['Free T4', 'T3', 'TSH']
cbc = ['WBC', 'RBC', 'Hemoglobin', 'Hematocrit', 'Platelet']
metabolic = ['T.Cholestero', 'Glu-FBS(S)', 'Insulin', 'HbA1c']
hormone = ['FSH', 'E2', 'Testosterone', 'LH']

✅ ❗ 왜 파일이 안 보이나요?
지금 상황은 오류가 아니라 환경의 제한 때문입니다.
제가 Python으로 파일을 생성하긴 했지만:
👉 그 파일이
사용자 PC(Excel)에 직접 저장된 것이 아니라
👉 임시 실행 환경(sandbox) 안에서 생성된 것입니다
그래서:
❌ 로컬 PC에서 파일 탐색기로는 안 보입니다
❌ 현재 Excel 파일에도 자동으로 붙지 않습니다

✅ ✅ 정확한 상태 설명
제가 생성한 파일들:

어린이코호트_임상데이터_Chat추가.xlsx
어린이코호트_임상데이터_ChatFullLog.xlsx
어린이코호트_자동분석리포트.xlsx

👉 이것들은 모두:
✔ 임시 분석 환경 안에서만 존재
✔ 실제 다운로드/저장은 별도 작업 필요

✅ ✅ 그럼 어떻게 해결하나요? (실무 방법)
🔥 방법 1 (추천): 직접 실행해서 저장
👉 가장 확실하고 재현 가능한 방법입니다
✅ STEP 1. 아래 코드 복사

In [20]:
import pandas as pd

file = '어린이코호트 임상데이터.xlsx'

df = pd.read_excel(file, sheet_name=0, engine='openpyxl')

# 기본 통계
desc = df.describe()

# 새 파일 저장
with pd.ExcelWriter('어린이코호트_자동분석리포트.xlsx', engine='openpyxl') as writer:
    df.to_excel(writer, sheet_name='RawData', index=False)
    desc.to_excel(writer, sheet_name='SummaryStats')

🔥 방법 2: 현재 Excel 파일에 직접 추가
👉 기존 파일을 유지하면서 시트만 추가

In [21]:
with pd.ExcelWriter(
    '어린이코호트 임상데이터.xlsx',
    engine='openpyxl',
    mode='a',
    if_sheet_exists='replace'
) as writer:
    
    desc.to_excel(writer, sheet_name='SummaryStats')

🔥 방법 3 (가장 많이 쓰는 방식)
👉 "분석용 파일 따로 만들기"
원본: 어린이코호트 임상데이터.xlsx
분석: 어린이코호트_Analysis.xlsx
로그: 어린이코호트_Log.xlsx

👉 이렇게 관리하면:

원본 손상 없음
기록 추적 쉬움
실무 표준 ✅

✅ ✅ 핵심 개념 (중요 🔥)

상황                의미
Copilot이 만든 파일  임시 환경
실제 Excel          로컬 PC
연결                ❌ 자동 연결 안됨

👉 그래서 항상:
✅ "코드 실행 → 로컬 저장" 이 필요합니다

✅ ✅ 앞으로 절대 잃어버리지 않는 방법
✔ 항상 파일 저장 위치 확인
✔ 코드 직접 실행
✔ append 방식 사용
✔ 파일 이름 버전 관리

🚀 ✅ 추천 (진짜 중요한 한 방)
이제부터는 이렇게 하세요:
👉 “분석 코드 파일(.py) + Excel” 같이 관리
예:
analysis.py
어린이코호트 임상데이터.xlsx

👉 실행할 때마다:

분석 결과 저장됨
로그 누적됨
100% 재현 가능

Create analysis.py

✅ Load your Excel
✅ Create analysis report
✅ Track ChatLog automatically
✅ Save everything persistently
✅ Append logs (no loss after logout)

✅ ✅ How to use (VERY IMPORTANT)

from terminal 


python analysis.py


🚀 Analysis started...
✅ Done!
📁 Report saved: 어린이코호트_자동분석리포트.xlsx
📊 Plots generated: ['ALT_hist.png', 'AST_hist.png', 'HbA1c_hist.png', 'TSH_hist.png']

✅ ✅ Optional upgrades (tell me if you want)
I can extend this into:
🔥 auto anomaly flagging (normal vs abnormal clinical)
🔥 PDF report (doctor-ready)
🔥 dashboard (Streamlit app)
🔥 1-click Excel button (VBA integration)

✅ ✅ Final takeaway
👉 You now have a system that:

✅ saves your chat
✅ tracks your analysis
✅ persists after logout
✅ is reusable forever

Below, I’ll cover all upgrades I suggested (and one extra best-practice option):

✅ Auto anomaly flagging (clinical “normal/abnormal” + rule-based categories)
✅ PDF report (doctor/manager-ready)
✅ Dashboard (Streamlit)
✅ One-click “double click to run” (EXE via PyInstaller)
✅ One-click run from Excel (button → run Python via xlwings)
⭐ Bonus: Office Scripts button (Excel web-friendly automation alternative)

0) Setup Once (recommended, keeps things reproducible)
0.1 Folder layout (create this)

cohort_project/
  analysis.py
  report_pdf.py
  app.py
  requirements.txt
  input/
    어린이코호트 임상데이터.xlsx
  output/
    (generated files here)
  logs/
    chatlog.csv

0.2 Create a virtual environment + install packages

cd cohort_project
python -m venv .venv
# Windows:
.venv\Scripts\activate
# macOS/Linux:
source .venv/bin/activate

pip install pandas openpyxl matplotlib reportlab streamlit pyinstaller xlwings

or pip install -r requirements.txt


``

1) Upgrade A — Auto Anomaly Flagging (Normal / Borderline / High)
What it does

Creates a sheet like Flags in your output workbook.
Adds columns such as:

TSH_flag, FreeT4_flag, HbA1c_flag, TChol_flag, FBS_flag


Uses your sheet’s reference criteria (e.g., TSH 0.6–4.8; FreeT4 0.9–1.67; HbA1c 4–6; and 3-level rules for Total Chol and FBS) [어린이코호트 임상데이터.xlsx | Excel]

2) Upgrade B — PDF Report (Charts + Key Tables)
What it does
Generates:

output/어린이코호트_Report.pdf
Includes:

Title page
Summary statistics table
Outlier counts
Embedded histograms (PNG) you already generate

ReportLab supports programmatic PDF creation via canvas and/or Platypus 

2.1 Create report_pdf.py

3) Upgrade C — Streamlit Dashboard (Interactive)
What it does

Launch a local web dashboard:

filters by age
show distributions
show outliers table
show correlation heatmap

Running locally is simply:

    streamlit run app.py

4) Upgrade D — One-click EXE (PyInstaller)
What it does
Builds a dist/analysis.exe (Windows) so teammates can run without typing commands.
PyInstaller one-file and build outputs are documented in official usage docs (--onefile, dist/, build/, .spec). [pyinstaller.org], [geeksforgeeks.org]
4.1 Build
From venv activated:

pyinstaller --onefile --name cohort_analysis analysis.py


In [ ]:
! pyinstaller --onefile --name cohort_analysis analysis.py
